src/util/rust_util/atomic_box.rs

```diff
@@ -0,0 +1,103 @@
+use std::sync::atomic::{AtomicPtr, Ordering};
+
+/// A lazily initialized box.  Similar to an `Option<Box<T>>`, but can be initialized atomically.
+///
+/// It is designed for implementing shared data.  Therefore, methods with `&self`, namely
+/// [`OnceOptionBox::get`] and the [`OnceOptionBox::get_or_init`] methods, only return shared
+/// references to the content (`&T`).  The user should use types that support multi-threaded
+/// accesses, such as mutexes or atomic types, if the inner type is supposed to be modified
+/// concurrently.
+///
+/// Once initialized, this object will own its content.  THe content is allocated in the heap, and
+/// will be dropped and deallocated when this instance is dropped.
+pub struct OnceOptionBox<T> {
```

Comment:
Can you explain why we need a new type like this, instead of something like `Once<Atomic<Option<T>>>` (T is the inner type, such as `usize` for `AtomicUsize`, or `MapState` for `Atomic<MapState>`)?

Reply:
I added comments that compare `OnceOptionBox` with `std::sync::OnceLock`